File: backend/app/models/product.py
import enum
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class SpecLabel(str, enum.Enum):
    RAM = "RAM"
    PROCESSOR = "Processor"
    GRAPHICS_CARD = "Graphics Card"
    STORAGE = "Storage"
    SCREEN = "Screen"


class Category(Base):
    __tablename__ = "categories"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    slug: Mapped[str] = mapped_column(String(120), unique=True, index=True, nullable=False)
    icon: Mapped[str | None] = mapped_column(String(50), nullable=True)  # lucide icon name
    parent_id: Mapped[int | None] = mapped_column(ForeignKey("categories.id"), nullable=True)

    parent: Mapped["Category | None"] = relationship(
        back_populates="children", remote_side="Category.id"
    )
    children: Mapped[list["Category"]] = relationship(
        back_populates="parent", cascade="all, delete-orphan"
    )

    products: Mapped[list["Product"]] = relationship(back_populates="category")


class Brand(Base):
    __tablename__ = "brands"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    logo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    products: Mapped[list["Product"]] = relationship(back_populates="brand")

class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    slug: Mapped[str] = mapped_column(String(255), unique=True, index=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    price: Mapped[float] = mapped_column(Float, nullable=False)
    original_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    badge: Mapped[str | None] = mapped_column(String(50), nullable=True)  # IN STOCK | CUSTOM BUILD
    stock: Mapped[int] = mapped_column(Integer, default=0)
    rating: Mapped[float] = mapped_column(Float, default=4.0)

    category_id: Mapped[int | None] = mapped_column(ForeignKey("categories.id"), nullable=True)
    brand_id: Mapped[int | None] = mapped_column(ForeignKey("brands.id"), nullable=True)

    category: Mapped["Category | None"] = relationship(back_populates="products")
    brand: Mapped["Brand | None"] = relationship(back_populates="products")

    images: Mapped[list["ProductImage"]] = relationship(
        back_populates="product",
        cascade="all, delete-orphan",
        order_by="ProductImage.position",
    )
    specs: Mapped[list["ProductSpec"]] = relationship(
        back_populates="product",
        cascade="all, delete-orphan",
    )

    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )


class ProductImage(Base):
    __tablename__ = "product_images"

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id"), nullable=False)
    image_url: Mapped[str] = mapped_column(String(500), nullable=False)
    is_primary: Mapped[bool] = mapped_column(Boolean, default=False)
    position: Mapped[int] = mapped_column(Integer, default=0)  # gallery order

    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    product: Mapped["Product"] = relationship(back_populates="images")


class ProductSpec(Base):
    __tablename__ = "product_specs"
    __table_args__ = (UniqueConstraint("product_id", "label", name="uq_product_spec_label"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id"), nullable=False)
    label: Mapped[SpecLabel] = mapped_column(Enum(SpecLabel), nullable=False)
    value: Mapped[str] = mapped_column(String(255), nullable=False)
    notes: Mapped[str | None] = mapped_column(String(255), nullable=True)

    product: Mapped["Product"] = relationship(back_populates="specs")

File: frontend/src/types/product.ts
// Mirrors app/schemas/product.py on the backend. Keep these in sync if the
// backend shapes change — there is no codegen step, this is hand-maintained.

export type SpecLabel = "RAM" | "Processor" | "Graphics Card" | "Storage" | "Screen";

export const SPEC_LABELS: SpecLabel[] = [
  "RAM",
  "Processor",
  "Graphics Card",
  "Storage",
  "Screen",
];

export interface ProductSpecIn {
  label: SpecLabel;
  value: string;
  notes?: string | null;
}

export interface ProductSpecOut extends ProductSpecIn {
  id: number;
}

export interface ProductImageOut {
  id: number;
  image_url: string;
  is_primary: boolean;
  position: number;
}

export interface ProductVariantOptionIn {
  group_label: string;
  option_label: string;
  image_url?: string | null;
  position?: number;
  is_default?: boolean;
}

export interface ProductVariantOptionOut extends ProductVariantOptionIn {
  id: number;
}

export interface CategoryOut {
  id: number;
  name: string;
  slug: string;
  icon: string | null;
  parent_id: number | null;
}

export interface CategoryTreeOut extends CategoryOut {
  children: CategoryTreeOut[];
}

export interface BrandOut {
  id: number;
  name: string;
}

export interface ProductOut {
  id: number;
  name: string;
  slug: string;
  description: string | null;
  price: number;
  original_price: number | null;
  badge: string | null;
  stock: number;
  rating: number;
  category_id: number | null;
  brand_id: number | null;
  category: CategoryOut | null;
  brand: BrandOut | null;
  images: ProductImageOut[];
  specs: ProductSpecOut[];
  variant_options: ProductVariantOptionOut[];
  created_at: string;
  updated_at: string;
}

export interface ProductListResponse {
  total: number;
  page: number;
  page_size: number;
  items: ProductOut[];
}

// ── Form-facing shapes (what the admin form builds before sending) ─────────

export interface ProductFormValues {
  name: string;
  description: string;
  price: string; // kept as string in the form, parsed to number on submit
  original_price: string;
  badge: string;
  stock: string;
  rating: string;
  category_id: number | null;
  slug: string;
  specs: ProductSpecIn[];
}

export interface ProductListFilters {
  page?: number;
  page_size?: number;
  search?: string;
  category_id?: number;
  brand_id?: number;
  min_price?: number;
  max_price?: number;
  sort?: "newest" | "price_asc" | "price_desc" | "name";
}